crypto/s2n_fips_rules.c

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "crypto/s2n_fips.h"
+#include "tls/s2n_cipher_suites.h"
+#include "tls/s2n_tls_parameters.h"
+#include "utils/s2n_result.h"
+
+/* FIPS requires at least 112 bits of security.
+ * https://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-107r1.pdf */
+const s2n_hash_algorithm fips_hash_algs[] = {
+    S2N_HASH_SHA224,
+    S2N_HASH_SHA256,
+    S2N_HASH_SHA384,
+    S2N_HASH_SHA512,
+};
+S2N_RESULT s2n_fips_validate_hash_algorithm(s2n_hash_algorithm hash_alg, bool *valid)
+{
+    RESULT_ENSURE_REF(valid);
+    *valid = false;
+    for (size_t i = 0; i < s2n_array_len(fips_hash_algs); i++) {
+        if (fips_hash_algs[i] == hash_alg) {
+            *valid = true;
+            return S2N_RESULT_OK;
+        }
+    }
+    return S2N_RESULT_OK;
+}
+
+/* https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-52r2.pdf */
+const uint8_t fips_cipher_suite_ianas[][2] = {
+    /* 3.3.1.1.1 Cipher Suites for ECDSA Certificates */
+    { TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256 },
+    { TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384 },
+    { TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256 },
+    { TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA384 },
+    { TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA },
+    { TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA },
+
+    /* 3.3.1.1.2 Cipher Suites for RSA Certificates */
+    { TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256 },
+    { TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384 },
+    { TLS_DHE_RSA_WITH_AES_128_GCM_SHA256 },
+    { TLS_DHE_RSA_WITH_AES_256_GCM_SHA384 },
+    { TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256 },
+    { TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA384 },
+    { TLS_DHE_RSA_WITH_AES_128_CBC_SHA256 },
+    { TLS_DHE_RSA_WITH_AES_256_CBC_SHA256 },
+    { TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA },
+    { TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA },
+    { TLS_DHE_RSA_WITH_AES_128_CBC_SHA },
+    { TLS_DHE_RSA_WITH_AES_256_CBC_SHA },
+
+    /* 3.3.1.1.2 3.3.1.2 Cipher Suites for TLS 1.3 */
+    { TLS_AES_128_GCM_SHA256 },
+    { TLS_AES_256_GCM_SHA384 },
+};
```

Comment:
The NIST guidelines sort of allow for RSA kx? Should we include that in the allowed things?
> While these guidelines do not recommend cipher suites using RSA key transport, there may be
circumstances in practice where RSA key transport is needed. For example, if an agency uses a
network appliance for regulatory or enterprise security purposes that only functions with these
cipher suites, then these cipher suites may need to be enabled.

NIST.SP.800-52r2 - Appendix D
I think I prefer the current decision (don't allow RSA kx ciphers) because it's much less ambiguous.

Reply:
It's an unfortunate complexity, but I think the two naming schemes makes sense. `S2N_FIPS_140_3` would be TLS 1.2+ and no RSA kx, and `S2N_FIPS_140_3_LEGACY_SUPPORT` would be TLS 1.0+ and allow RSA kx?

Reply:
RSA key exchange is disallowed by NIST at the end of 2023. So, I don't think we should include any `TLS_RSA_*` ciphersuites in the FIPS policies.